zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/MigrationSnapshotDirector.java

```diff
@@ -0,0 +1,204 @@
+  }
+
+  @Override
+  public void removeFailureListener(final FailureListener failureListener) {
+    listeners.remove(failureListener);
+  }
+
+  private void forceSnapshotUntilSuccessful() {
+    forceSnapshotUntilSuccessful(null);
+  }
+
+  private void forceSnapshotUntilSuccessful(final Throwable error) {
+    if (!isRecoverableError(error)) {
+      LOG.warn("Snapshot cannot be taken due to unrecoverable error", error);
+      close();
+      return;
+    }
+    if (!snapshotTaken && runningSnapshot == null) {
+      final var nextDelay = retryState.nextDelay();
+      final var issue =
+          error == null
+              ? HealthIssue.of(
+                  String.format(
+                      "Snapshot not taken yet, retryCount=%d, will retry in %s",
+                      retryState.getRetryCount(), nextDelay),
+                  Instant.now())
+              : HealthIssue.of(error, Instant.now());
+      healthReport = healthReport.withIssue(issue);
+      notifyListeners();
+      runningSnapshot =
+          control.schedule(
+              nextDelay,
+              () -> {
+                forceSnapshot()
+                    .onComplete(
+                        (ignored, ex) -> {
+                          forceSnapshotUntilSuccessful(ex);
+                        },
+                        control);
+              });
+    }
+  }
+
+  private boolean isRecoverableError(final Throwable error) {
+    return error == null
+        || (error instanceof IOException
+            || (error instanceof CompletionException && error.getCause() instanceof IOException));
+  }
+
+  private ActorFuture<Void> forceSnapshot() {
+    if (!snapshotTaken) {
+      return snapshotDirector
+          .forceSnapshot()
+          .andThen(
+              (snapshot, error) -> {
+                if (snapshot != null) {
+                  // reset the flag
+                  snapshotTaken = true;
+                  LOG.debug("Snapshot taken after migrations: {}", snapshot.getId());
+                  healthReport = HealthReport.healthy(this);
+                  notifyListeners();
+                }
+                runningSnapshot = null;
+                return error != null
+                    ? CompletableActorFuture.completedExceptionally(error)
+                    : CompletableActorFuture.completed(null);
+              },
+              control);
+    } else {
+      return CompletableActorFuture.completed(null);
+    }
+  }
+
+  private void notifyListeners() {
+    listeners.forEach(
+        (listener, ignored) -> {
+          final var report = getHealthReport();
+          listener.onHealthReport(report);
+        });
+  }
+
+  private HealthReport snapshotNotTaken() {
+    return HealthReport.unhealthy(this).withMessage("No snapshot taken yet", Instant.now());
+  }
+
+  private static final class RetryState {
+    private int retryCount = 0;
+    private final ExponentialBackoff backoff = new ExponentialBackoff(5000, 500);
```

Comment:
🔧 I just realized we can replace this whole `RetryState` with `ExponentialBackoffRetryDelay` 😬 

Ah well. You don't need to do this here right now.